### DISCUSSION TITLE
Support Garage, a more lightweight S3 server

### DISCUSSION THREAD

**Discussion by helmut72** [outline/outline#5432] [Ideas]
Would be cool if Outline works with a more lightweight S3 server like [Garage](https://garagehq.deuxfleurs.fr).

Thank you.

**Comment by tommoor** [comment]
Have you tried it? Is there something specifically that doesn’t work?

**Comment by helmut72** [comment]
It looks like Outline need anonymous access to S3, but Garage doesn't support it:
https://git.deuxfleurs.fr/Deuxfleurs/garage/issues/263

From this issue I understand, anonymous access is only possible as workaround, when exposing a bucket as a website:
https://garagehq.deuxfleurs.fr/documentation/cookbook/exposing-websites/

Using different subdomains (write.s3.domain.com for writing, read.s3.domain.com for reading) with a reverse proxy works for me, but Outline doesn't support different subdomains for writing and reading.

**Reply by tommoor** [reply]
Gotcha, should be easy enough - folks do have a lot of difficulties with minio so it would be nice to have a simpler option if possible

**Reply by 0x5c** [reply]
What area of the code would be affected? I'm interested in trying out outline on garage, and might want to have a got at implementing what's needed

**Reply by tommoor** [reply]
Currently there is an exception – when the `AWS_S3_ACL` is set to `private` profile pictures are still uploaded with public permissions. We need to:

- Remove that exception
- Ensure that the `attachments.redirect` endpoint can be called without authentication and allows redirecting to these "public" resources that are stored in a private bucket

**Reply by 0x5c** [reply]
Thanks for the pointers!
I did some scouting, it seems that the ACL is set for both `Upload{User,Team}AvatarTask` by these params to `uploadToS3FromUrl` https://github.com/outline/outline/blob/9ef375d83cfa13a1c2addeed740e024bdd84c203/server/queues/tasks/UploadUserAvatarTask.ts#L26
https://github.com/outline/outline/blob/9ef375d83cfa13a1c2addeed740e024bdd84c203/server/queues/tasks/UploadTeamAvatarTask.ts#L26
I assume the hardcoded ACL needs to be removed from both.

It also seems that `AttachmentHelper.presetToAcl()` has a special-case to set that ACL, but from what I understand `AttachmentHelper` is not currently in use for avatar upload, is that correct? In this case, should anything be done with it?
https://github.com/outline/outline/blob/9ef375d83cfa13a1c2addeed740e024bdd84c203/server/models/helpers/AttachmentHelper.ts#L36-L43

As for the `attachments.redirect` endpoint, I'm still confused as to what it currently can and cannot do, and what needs to be done. I'm gonna try taking care of the upload part and see if things work or not from that point.

**Reply by tommoor** [reply]
> I assume the hardcoded ACL needs to be removed from both.

Yep

> from what I understand AttachmentHelper is not currently in use for avatar upload

We could use it

> As for the attachments.redirect endpoint, I'm still confused as to what it currently can and cannot do, and what needs to be done

The auth needs to be made optional, auth middleware takes options to achieve this. You then need to ensure the auth is present and the correct team for private attachments

**Reply by 0x5c** [reply]
Current status is that I'm blocked on setting up test instance since I can't find any easy to setup authentication provider that works locally, do you have a recommendation?

**Reply by tommoor** [reply]
There is a seed script for this purpose,

```
yarn build:server && node build/server/scripts/seed.js my@email.com
```

**Reply by tommoor** [reply]
You can then sign-in via email, the magic-link is printed in the console in development

**Comment by helmut72** [comment]
For me this feature request can be closed, because your new local filesystem is a better solution.
